File: arm.64/langs/node/src/bind.ts
import path from 'node:path';
import koffi from 'koffi';
import { Cabi, DefStruct, MoreFn, PumpFn, LessFn } from './type.js';

export function bind(binaryPath: string): Cabi | null {
  const filename = path.basename(binaryPath);
  if (['c', 'cpp', 'rust', 'go', 'swift', 'haskell', 'zig', 'v', 'slint_sample'].includes(filename)) {
    return null;
  }

  try {
    const lib = koffi.load(binaryPath);

    const More = lib.func('More', 'int', [koffi.pointer(DefStruct)]) as MoreFn;
    const Pump = lib.func('Pump', 'int', ['str', 'str', 'str']) as PumpFn;
    const Less = lib.func('Less', 'int', [koffi.pointer(DefStruct)]) as LessFn;

    return {
      name: filename,
      path: binaryPath,
      lib,
      More,
      Pump,
      Less,
    };
  } catch {
    return null;
  }
}

File: arm.64/langs/rust/src/bind.rs
use crate::r#type::{Cabi, LessFn, MoreFn, PumpFn};
use libloading::{Library, Symbol};
use std::path::Path;

pub fn bind(binary_path: &str) -> Option<Cabi> {
    let filename = Path::new(binary_path).file_name()?.to_str()?;
    if matches!(filename, "c" | "cpp" | "rust" | "go" | "swift" | "haskell" | "zig" | "v" | "slint_sample") {
        return None;
    }

    unsafe {
        let lib = Library::new(binary_path).ok()?;
        let more: Option<MoreFn> = lib.get(b"More\0").ok().map(|s: Symbol<MoreFn>| *s);
        let pump: Symbol<PumpFn> = lib.get(b"Pump\0").ok()?;
        let pump_fn: PumpFn = *pump;
        let less: Option<LessFn> = lib.get(b"Less\0").ok().map(|s: Symbol<LessFn>| *s);

        Some(Cabi {
            name: filename.to_string(),
            path: binary_path.to_string(),
            lib,
            more,
            pump: pump_fn,
            less,
        })
    }
}
